packages/teams-js/src/public/app.ts

```diff
@@ -806,6 +816,100 @@ export namespace app {
       resolve(sendAndHandleStatusAndReason('executeDeepLink', deepLink));
     });
   }
+
+  /**
+   * Namespace to opt into full-caching or delayed-termination of an app.
+   *
+   * @beta
+   */
+  export namespace lifecycle {
+    /**
+     * Register on resume handler function type
+     *
+     * @param context - Data structure to be used to pass the context to the app.
+     */
+    type registerOnResumeHandlerFunctionType = (context: ResumeContext) => void;
+    /**
+     * A function that is called when the app is ready to be suspended or terminated.
+     */
+    type ReadyToSuspendOrTerminateFunction = () => void;
+
+    /**
+     * Register before suspendOrTerminate handler function type
+     *
+     * @param readyToSuspendOrTerminate - A function that the handler is called with. Apps should call this function when they are ready.
+     * @returns void
+     */
+    type registerBeforeSuspendOrTerminateHandlerFunctionType = (
+      readyToSuspendOrTerminate: ReadyToSuspendOrTerminateFunction,
+    ) => void;
+
+    /**
+     * Registers a handler to be called before the page is suspended or terminatd.
+     *
+     * @param handler - The handler to invoke before the page is suspended or terminated. The page should
+     * invoke the readyToSuspendOrTerminate function provided to it once it's ready.
+     *
+     */
+    export function registerBeforeSuspendOrTerminateHandler(
+      handler: registerBeforeSuspendOrTerminateHandlerFunctionType,
+    ): void {
+      handler && ensureInitialized(runtime);
+      if (handler && !isSupported()) {
+        throw errorNotSupportedOnPlatform;
+      }
+      Handlers.registerBeforeSuspendOrTerminateHandler(handler);
+    }
+
+    /**
+     * Checks if app.lifecycle is supported by the host.
+     * @returns boolean to represent whether the lifecycle capability is supported
+     * @throws Error if {@linkcode app.initialize} has not successfully completed
+     *
+     * @beta
+     */
+    export function isSupported(): boolean {
+      return ensureInitialized(runtime) && runtime.supports.app && runtime.supports.app.lifecycle ? true : false;
+    }
+
+    /**
+     * Namespace to opt into full caching of an app
+     *
+     * @beta
+     */
+    export namespace caching {
+      /**
+       * Registers a handler to be called when the page has been requested to load.
+       *
+       * @param handler - The handler to invoke when the page is loaded.
+       *
+       * @beta
+       */
+      export function registerOnResumeHandler(handler: registerOnResumeHandlerFunctionType): void {
```

Comment:
>registerOnResumeHandler [](http://example.com/codeflow?start=22&length=23)

Maybe we talked about this as part of the dev plan and I just forgot. From the perspective of the developer API, is there value in separating out this registration function into its own subcapability? Yes, there will be hosts which don't fully support the suspend/resume lifecycle, but since the only function here is the registration, and registering a handler in a hub which doesn't support it will just result it in the handler never getting called, can this just go directly into the `lifecycle` capability?

Reply:
We divided caching into two: delayed termination and full caching. Hosts can choose to support only delayed termination (In web, they don't have full caching option) or full caching.
To trace that behavior to teams-js, we have 'lifecycle' that indicates delayed termination and subcapability 'caching' that is full caching. 
In teams-js, if we have both handlers in one capability and a host is supporting only delayed termination, what should isSupported function return: true or false? 
If we return true, app developer might get confused, because app is not being cached, while capability shows supported. 

However, you can also argue that app might not get cached for other reasons like memory limitation etc even though  hub supports full caching, and is-supported for subcapability 'caching' returns true. 

Last time in our caching sync, the votes for subcapability were higher. If you think nested subacapility is causing more harm than benefits, we are open to change it to one capability :)

Reply:
First, I don't want to be in the business of overruling democratically elected capability names.

I will state my position and y'all, but do treat it as just one input.

Because caching is an optimization, no developer should ever be relying on it. Even in hosts where caching exists, as you say, it might not ever happen for a variety of reasons. As such, the `isSupported` method on the `caching` namespace (as you currently have it) is actually not helpful since it doesn't guarantee anything (other than `registerOnResumeHandler` won't throw an error). Since the act of registering a handler is a supported operation on all hosts, and there's no host where you can guarantee that it will be called, I didn't see the reason to increase developer and code complexity by putting it into its own subcapability, at least from the teamsjs perspective, since they aren't getting any value from that.

Reply:
Thank you for more feedback! We decided to combine both under one subCapability. The isSupported function will return true if hubs provide lifeCycleService.